rust/private/rust.bzl

```diff
@@ -118,6 +118,48 @@ def crate_root_src(attr, srcs, crate_type):
         fail("No {} source file found.".format(" or ".join(file_names)), "srcs")
     return crate_root
 
+def _invalid_chars_in_crate_name(name):
+    """Returns any invalid chars in the given crate name.
+
+    Args:
+        name (str): Name to test.
+
+    Returns:
+        list: List of invalid characters in the crate name.
+    """
+
+    return dict([(c, ()) for c in name.elems() if not (c.isalnum() or c == "_")]).keys()
+
+def _crate_name(label, attr):
+    """Returns the crate name to use for the current target.
+
+    Args:
+        label (Label): The label of the current target.
+        attr (struct): The attributes of the current target.
+
+    Returns:
+        str: The crate name to use for this target.
+    """
+    if attr.crate_name:
```

Comment:
Can we unify these codepaths? Maybe with a prefix of something like:
```python
crate_name, crate_name_source = attr.name, "name"
if attr.crate_name:
    crate_name = attr.crate_name
    crate_name_source = "crate_name"
...
```

Reply:
I've tried this, but I'm not sure it's really better. Here's what I've got:

```
    crate_name, crate_name_source = name_to_crate_name(attr.name), "name"
    if attr.crate_name:
      crate_name = attr.crate_name
      crate_name_source = "crate_name"

    invalid_chars = _invalid_chars_in_crate_name(crate_name)

    if invalid_chars:
        if crate_name_source == "name":
            fail(
                "Crate name '{}'".format(crate_name),
                "derived from Bazel target name '{}'".format(label.name),
                "contains invalid character(s): {}".format(" ".join(invalid_chars)),
                "\nConsider adding a crate_name attribute to set a valid crate name",
            )
        else:
            fail("Crate name '{}' contains invalid character(s): {}".format(
                attr.crate_name,
                " ".join(invalid_chars),
            ))

    return crate_name
```

This is more lines than what I have above and uses "magic" strings for `crate_name_source` (which have the potential for typos).

Side note: I see you using `attr.name` instead of `label.name`. Is there a preference for one or the other in Bazel rules? I do see that in this particular case, using `attr.name` would allow me to eliminate the `label` parameter entirely, which seems attractive, but I wanted to ask first.

Reply:
I believe `ctx.attr.name` and `ctx.label.name` are identical. I'd default to using `ctx.attr.name` but I don't have a better reason than that I think it's more common.

Reply:
I was imagining something like:

```python
crate_name, source = attr.crate_name, "name"
if attr.crate_name:
    crate_name = attr.crate_name
    crate_name_source = "crate_name"

invalid_chars = _invalid_chars_in_crate_name(crate_name)

if invalid_chars:
    fail(
        "Crate name {crate_name} {source_description} contains invalid character(s): {invalid_characters}.{suggestion}".format(
            crate_name = crate_name,
            invalid_characters = " ".join(invalid_chars),
            source_description = "derived from Bazel target name {}".format(label.name) if crate_name_source == "name" else "from crate_name attribute",
            suggestion = "\nConsider adding a crate_name attribute to set a valid crate name" if crate_name_source == "crate_name" else "",
        )
    )
return crate_name
```

But I don't think either way is clearly better :) Happy to merge either way!

Reply:
Ah, got it. If you're fine with either, I'd prefer to keep the original version -- some duplication, but less logic.

Reply:
> I believe `ctx.attr.name` and `ctx.label.name` are identical. I'd default to using `ctx.attr.name` but I don't have a better reason than that I think it's more common.

I'm going with `ctx.attr.name` then because it allows me to eliminate the `label` parameter.